README.md
# CheckServerConnections
<h1>Overview</h1>
This is a gui created with tkinter for managing a program that checks connections to a list of servers or workstations.
Input a list of servers you would like to keep track of. The gui will insert them into a sqllite database for persistent storage
Run the Gui.py file and it will bring you to the first page(image below). 
</br>
</br>
After this click on the second page to add a file location(bottom image). Write any file location. By default the file will appear in same folder as 
the program.
<br/>
<br/>
At this point press Run Scan and the application will scan the list of servers and write the results to the file locations listed.
Example of the format below and in history.txt/ newlogs file in repo.
</br>
![scanres](https://user-images.githubusercontent.com/63485111/147626883-63a4a258-bfd6-41ce-9437-2fd38335fa0b.PNG)

<br/>
<br/>
<h2>Optional Gmail configuration</h2>
There is an optional feature that if a during a scan a high priority server is down the program will automatically email you with a message containing details.
<p>In the Gmail.py on lines 12 & 13 enter gmail user and password. For this to work you will need to change a setting in gmail that allows less secure apps</p>
<p>This is completely optional and the application works fully without the additional email configuration</p>
</br>
<h2>Additional use information</h2>
Once you configure the output locations and server list with the gui. You can schedule cronjob for python to run CheckServer.py.
The scans will be automated to run at custom time. If you configure email update you won't even need to check logs because you'll get emailed if a server is down




<br/>
<br/>
<br/>




![server_conn_gui1](https://user-images.githubusercontent.com/63485111/147626095-28c3de3d-e20f-40cd-850d-b6629d9a2ce6.PNG)


![server_conn_gui2](https://user-images.githubusercontent.com/63485111/147626132-db17ac2c-37d7-4719-b7d4-981052a5e52b.PNG)


history.txt
amazon.com is up on port 443 with ssl
Time: 2021-12-28 21:51:34.249734
Success Status: True
___________________________
__

onycalo.com is up on port 80 with plain
Time: 2021-12-27 22:22:15.902219
Success Status: True
___________________________
calochess.com is up on port 80 with plain
Time: 2021-12-27 22:22:15.942605
Success Status: True
___________________________
192.168.1.18 is not reachable
Time: 2021-12-27 22:22:16.016552
Success Status: False
___________________________
reddit.com is up on port 80 with plain
Time: 2021-12-27 22:24:22.790658
Success Status: True
___________________________
anthonycalo.com is up on port 443 with ssl
Time: 2021-12-27 22:24:22.830277
Success Status: True
___________________________
calochess.com is up on port 80 with plain
Time: 2021-12-27 22:24:22.956351
Success Status: True
___________________________
192.168.1.18 is not reachable
Time: 2021-12-27 22:24:23.000224
Success Status: False
___________________________
reddit.com is up on port 80 with plain
Time: 2021-12-27 22:30:52.658366
Success Status: True
___________________________
anthonycalo.com is up on port 443 with ssl
Time: 2021-12-27 22:30:52.706593
Success Status: True
___________________________
calochess.com is up on port 80 with plain
Time: 2021-12-27 22:30:52.771040
Success Status: True
___________________________
192.168.1.18 is not reachable
Time: 2021-12-27 22:30:52.835052
Success Status: False
___________________________
ac-encryption.com is up on port 443 with ssl
Time: 2021-12-27 22:30:52.841557
Success Status: True
___________________________
reddit.com is up on port 80 with plain
Time: 2021-12-27 22:32:50.289354
Success Status: True
___________________________
anthonycalo.com is up on port 443 with ssl
Time: 2021-12-27 22:32:50.336819
Success Status: True
___________________________
calochess.com is up on port 80 with plain
Time: 2021-12-27 22:32:50.426973
Success Status: True
___________________________
192.168.1.18 is not reachable
Time: 2021-12-27 22:32:50.466346
Success Status: False
___________________________
anthonycalo.com is up on port 443 with ssl
Time: 2021-12-28 14:07:48.001890
Success Status: True
___________________________
calochess.com is up on port 80 with plain
Time: 2021-12-28 14:07:48.156546
Success Status: True
___________________________
anthonycalo.com is up on port 443 with ssl
Time: 2021-12-28 16:14:34.202273
Success Status: True
___________________________
calochess.com is up on port 80 with plain
Time: 2021-12-28 16:14:34.373083
Success Status: True
___________________________
192.168.1.18 is not reachable
Time: 2021-12-28 16:14:34.488095
Success Status: False
___________________________
anthonycalo.com is up on port 443 with ssl
Time: 2021-12-28 16:20:53.443351
Success Status: True
___________________________
calochess.com is up on port 80 with plain
Time: 2021-12-28 16:20:53.570938
Success Status: True
___________________________
Server had this error: [WinError 10054] An existing connection was forcibly closed by the remote host
Time: 2021-12-28 16:20:53.635531
Success Status: False
___________________________
anthonycalo.com is up on port 443 with ssl
Time: 2021-12-28 16:21:19.570613
Success Status: True
___________________________
calochess.com is up on port 80 with plain
Time: 2021-12-28 16:21:19.718848
Success Status: True
___________________________
amazon.com is up on port 443 with ssl
Time: 2021-12-28 16:21:19.779901
Success Status: True
___________________________
anthonycalo.com is up on port 443 with ssl
Time: 2021-12-28 16:56:50.278464
Success Status: True
___________________________
calochess.com is up on port 80 with plain
Time: 2021-12-28 16:56:50.386437
Success Status: True
___________________________
amazon.com is up on port 443 with ssl
Time: 2021-12-28 16:56:50.449013
Success Status: True
___________________________
Server had this error: [SSL: WRONG_VERSION_NUMBER] wrong version number (_ssl.c:1076)
Time: 2021-12-28 16:56:50.549894
Success Status: False
___________________________
anthonycalo.com is up on port 443 with ssl
Time: 2021-12-28 16:59:12.550299
Success Status: True
___________________________
calochess.com is up on port 80 with plain
Time: 2021-12-28 16:59:12.690526
Success Status: True
___________________________
amazon.com is up on port 443 with ssl
Time: 2021-12-28 16:59:12.737124
Success Status: True
___________________________
Server had this error: [SSL: WRONG_VERSION_NUMBER] wrong version number (_ssl.c:1076)
Time: 2021-12-28 16:59:12.836120
Success Status: False
___________________________
anthonycalo.com is up on port 443 with ssl
Time: 2021-12-28 17:02:57.584739
Success Status: True
___________________________
calochess.com is up on port 80 with plain
Time: 2021-12-28 17:02:57.649469
Success Status: True
___________________________
amazon.com is up on port 443 with ssl
Time: 2021-12-28 17:02:57.723976
Success Status: True
___________________________
anthonycalo.com is up on port 443 with ssl
Time: 2021-12-28 21:52:31.270515
Success Status: True
___________________________
calochess.com is up on port 80 with plain
Time: 2021-12-28 21:52:31.375352
Success Status: True
___________________________
amazon.com is up on port 443 with ssl
Time: 2021-12-28 21:52:31.500814
Success Status: True
___________________________
anthonycalo.com is up on port 443 with ssl
Time: 2021-12-28 21:53:05.112742
Success Status: True
___________________________
calochess.com is up on port 80 with plain
Time: 2021-12-28 21:53:05.203199
Success Status: True
___________________________
apptracker.us is up on port 80 with plain
Time: 2021-12-28 21:53:05.256630
Success Status: True
___________________________
anthonycalo.com is up on port 443 with ssl
Time: 2021-12-28 21:56:35.125007
Success Status: True
___________________________

Gui.py
from tkinter import *
import tkinter.font as font
import backend_db as backend
import CheckServer
from tkinter import ttk
root =Tk()
backend.connect()

#functions/db functions_______________________________________
def run_scan():
    serverList=[]
    for row in backend.view_servers():
        #minor difference is that 3, 2 are swapped because connnection and port are in diff orders in db vs
        #server class
        serverList.append(CheckServer.Server(row[1],row[3],row[2],row[4]))
    for server in serverList:
        server.check_connection()
def get_selected_row(event):
    global selected_record
    try:
        index=list1.curselection()[0]
        selected_record=list1.get(index)
        #print(type(selected_record[1]))
        

        E1.delete(0, END)
        E1.insert(END, selected_record[1])
        # E2.delete(0, END)
        # E2.insert(END, selected_record[2])
        connection_text.set(selected_record[2])
        E3.delete(0, END)
        E3.insert(END, selected_record[3])
        priority_text.set(selected_record[4])
        return selected_record[0]
    except:
        pass

def delete_command():
    backend.delete(selected_record[0])

def view_command():
    list1.delete(0, END)
    for row in backend.view_servers():
        list1.insert(END, row)

def search_command():
    list1.delete(0, END)
    for row in backend.search(URL_text.get(), connection_text.get(), port_text.get(), priority_text.get()):
        list1.insert(END, row)

def insert_command():
    list1.delete(0, END)
    print(priority_text)
    print(priority_text.get(), "<-.current")
    backend.insert(URL_text.get(), connection_text.get(), port_text.get(), priority_text.get())

def update_command():
    backend.update(selected_record[0], URL_text.get(), connection_text.get(), port_text.get(), priority_text.get())

def page2_command():
    root.destroy()
    import Gui_p2

#hover command
def changeOnHover(button, colorOnHover, colorOnLeave):
  
    # adjusting backgroung of the widget
    # background on entering widget
    button.bind("<Enter>", func=lambda e: button.config(
        background=colorOnHover,
        fg='white'
        ))
  
    # background color on leving widget
    button.bind("<Leave>", func=lambda e: button.config(
        background=colorOnLeave))


#Gui section _____________________________________________________________
buttonFont=font.Font(size="15", family="Arial")
dropFont=font.Font(size="20", family="Helvectica")

L1 = Label(root, text="URL", font=("Courier", 25), bg="#B8894A")
L1.grid(row=1, column=2, padx=25)

L2 = Label(root, text="Connection Type", font=("Courier", 25), bg="#B8894A")
L2.grid(row=3, column=2, padx=25)

L3 = Label(root, text="Port", font=("Courier", 25), bg="#B8894A")
L3.grid(row=1, column=6, padx=50)

L4 = Label(root, text="Priority", font=("Courier", 25), bg="#B8894A")
L4.grid(row=3, column=6, padx=50)

#entrys
URL_text=StringVar()
E1=Entry(root, textvariable=URL_text, font=("Helvectica", 20))
E1.grid(row=2, column=2, padx=25)

connection_text=StringVar()
connection_list=("plain", "ssl", "ping", " ")

# E2=Entry(root, textvariable=connection_text, font=("Helvectica", 20))
# E2.grid(row=4, column=2, padx=25)
connection_text.set("plain")
e2_dropdown=ttk.Combobox(root, values=connection_list, textvariable=connection_text)
e2_dropdown.set(connection_text)
e2_dropdown.config(width=18, height=1, font=dropFont)
e2_dropdown['state']="readonly"
e2_dropdown.grid(row=4, column=2) 


port_text=StringVar()
E3=Entry(root, textvariable=port_text, font=("Helvectica", 20))
E3.grid(row=2, column=6, padx=25)


priority_text=StringVar()
priority_list=("high", "low")
# E4=Entry(root, textvariable=priority_text, font=("Helvectica", 20))
# E4.grid(row=4, column=6, padx=25)
e4_dropdown=ttk.Combobox(root, values=priority_list, textvariable=priority_text)
e4_dropdown.set(priority_text)
e4_dropdown.config(width=18, height=1, font=dropFont)
e4_dropdown['state']="readonly"
e4_dropdown.grid(row=4, column=6) 

list1 = Listbox(root, height=15, width=30,font=("Helvectica", 20))
list1.grid(row=5, column=2, rowspan=6, columnspan=5,padx=25, pady=50)

list1.bind('<<ListboxSelect>>', get_selected_row)


b1=Button(root, text="View all", width=12, height=2,font=buttonFont, bg="blue",fg="white",command=view_command)
b1.grid(row=5, column=7)

b2=Button(root, text="Search Entry", width=12, height=2, font=buttonFont, bg="blue",fg="white", command=search_command)
b2.grid(row=6, column=7)

b3=Button(root, text="Add Entry", width=12, height=2, font=buttonFont, bg="blue",fg="white", command=insert_command)
b3.grid(row=7, column=7, pady=0)

b4=Button(root, text="Update select", width=12, height=2, font=buttonFont, bg="blue",fg="white", command=update_command)
b4.grid(row=8, column=7)

b5=Button(root, text="Delete Selected", width=12, height=2, font=buttonFont, bg="blue",fg="white", command=delete_command)
b5.grid(row=9, column=7)

b6=Button(root, text="Close",height=2, width=12,font=buttonFont, bg="blue",fg="white", command=root.destroy)
b6.grid(row=10, column=7)

b7=Button(root, text="Run Scan", height=2, width=25,font=buttonFont, bg="blue",fg="white",  command=run_scan)
b7.grid(row=11, column=2)

b8=Button(root, text="Change output location", height=2, width=25,font=buttonFont, bg="blue",fg="white",  command=page2_command)
b8.grid(row=11, column=6)

changeOnHover(b1, "green", "blue")
changeOnHover(b2, "green", "blue")
changeOnHover(b3, "green", "blue")
changeOnHover(b4, "green", "blue")
changeOnHover(b5, "green", "blue")
changeOnHover(b5, "green", "blue")
changeOnHover(b6, "green", "blue")
changeOnHover(b7, "green", "blue")
changeOnHover(b8, "green", "blue")

root.title("Server Scanner")
root.geometry("900x850")
root.configure(bg="#B8894A")
root.mainloop()


Gui_p2.py
from tkinter import *
import tkinter.font as font
import backend_db as backend

root=Tk()

#functions/db functions_______________________________________
def get_selected_row(event):
    global selected_record
    try:
        index=list1.curselection()[0]
        selected_record=list1.get(index)
        #print(type(selected_record[1]))
        

        E1.delete(0, END)
        E1.insert(END, selected_record[1])

        return selected_record[0]
    except:
        pass
def view_command():
    list1.delete(0, END)
    for row in backend.view_locations():
        list1.insert(END, row)

def insert_command():
    list1.delete(0, END)
    print(location_text.get())
    backend.insert_location(location_text.get())

def delete_command():
    backend.delete_location(selected_record[0])

def page1_command():
    root.destroy()
    import Gui

#Gui section _____________________________________________________________
buttonFont=font.Font(size="15", family="Arial")
L1 = Label(root, text="Log Location", font=("Courier", 25), bg="#B8894A")
L1.grid(row=1, column=2, ipadx=30)

#entrys
location_text=StringVar()
E1=Entry(root, textvariable=location_text, font=("Helvectica", 20))
E1.grid(row=2, column=2, ipadx=30)

#buttons___________________________________________________________________
b1=Button(root, text="View all", width=12, height=2,font=buttonFont, bg="blue",fg="white",command=view_command)
b1.grid(row=4, column=3)

b2=Button(root, text="Add Entry", width=12, height=2, font=buttonFont, bg="blue",fg="white", command=insert_command)
b2.grid(row=5, column=3, pady=0)

delete_button=Button(root, text="Delete Selected", width=12, height=2, font=buttonFont, bg="blue",fg="white", command=delete_command)
delete_button.grid(row=6, column=3)

b6=Button(root, text="Close",height=2, width=12,font=buttonFont, bg="blue",fg="white", command=root.destroy)
b6.grid(row=7, column=3)

infoFont=font.Font(size="15", family="Arial", weight="bold")
l2=Label(root, text="* this scan results will be written to locations in list", fg="red",bg="#B8894A")
l2.grid(row=8, column=2)
l2.config(font=infoFont)

page1=Button(root, text="Go to main page", height=2, width=25,font=buttonFont, bg="blue",fg="white",  command=page1_command)
page1.grid(row=9, column=2)


#_________________________________________________________________________________________________

list1 = Listbox(root, height=10, width=30,font=("Helvectica", 20))
list1.grid(row=4, column=2, rowspan=5, padx=25, pady=50)

list1.bind('<<ListboxSelect>>', get_selected_row)

root.title("Server Scanner")
root.geometry("900x850")
root.configure(bg="#B8894A")
root.mainloop()

backend_db.py
import sqlite3

def connect():
    conn=sqlite3.connect("servers.db")
    cur=conn.cursor()
    cur.execute("CREATE TABLE IF NOT EXISTS servers (id INTEGER PRIMARY KEY, url text, connection text, port integer, priority integer)")
    cur.execute("CREATE TABLE IF NOT EXISTS outputs (id INTEGER PRIMARY KEY, location text)")
    conn.commit()
    conn.close()

def insert(url, connection, port, priority):
    conn=sqlite3.connect("servers.db")
    cur=conn.cursor()
    cur.execute("INSERT INTO servers VALUES (NULL,?, ?, ?,?)",(url, connection, port, priority))
    conn.commit()
    conn.close()

def view_locations():
    conn=sqlite3.connect("servers.db")
    cur=conn.cursor()
    cur.execute("SELECT * FROM outputs")
    rows=cur.fetchall()
    conn.close()
    return(rows)

def insert_location(location):
    conn=sqlite3.connect("servers.db")
    print(location)
    cur=conn.cursor()
    cur.execute("INSERT INTO outputs VALUES (NULL, ?)", ((location,)))
    conn.commit()
    conn.close()

def delete_location(id):
    conn=sqlite3.connect("servers.db")
    cur=conn.cursor()
    cur.execute("DELETE FROM outputs WHERE id={}".format(id))
    conn.commit()
    conn.close()

def view_servers():
    conn=sqlite3.connect("servers.db")
    cur=conn.cursor()
    cur.execute("SELECT * FROM servers")
    rows=cur.fetchall()
    conn.close()
    return(rows)

def search(url="", connection="", port="", priority=""):
    conn=sqlite3.connect("servers.db")
    cur=conn.cursor()
    cur.execute("SELECT * FROM servers where url=? OR connection=? OR port=? OR priority=?",(url, connection, port, priority))
    rows=cur.fetchall()
    conn.close()
    return rows

def delete(id):
    conn=sqlite3.connect("servers.db")
    cur=conn.cursor()
    cur.execute("DELETE FROM servers WHERE id={}".format(id))
    conn.commit()
    conn.close()

def update(id, url, connection, port, priority):
    conn=sqlite3.connect("servers.db")
    cur=conn.cursor()
    cur.execute("UPDATE servers SET url=?, connection=?, port=?, priority=? WHERE id=?",(url, connection, port, priority, id))
    conn.commit()
    conn.close()

CheckServer.py
import socket
import ssl
from datetime import datetime 
import requests
import Gmail
import backend_db as backend

class Server():
    def __init__(self, name, port, connection, priority):
        self.name=name
        self.port=port
        self.connection=connection.lower()
        self.priority=priority.lower()
        self.alert=False
        self.locations=backend.view_locations()
    
    def check_connection(self):
        msg=""
        success=False
        now=datetime.now()

        try:
            if(self.connection=="plain"):
                socket.create_connection((self.name, self.port), timeout=8)
                msg="{} is up on port {} with {}".format(self.name, self.port, self.connection)
                success=True
            elif (self.connection=="ssl"):
                ssl.wrap_socket(socket.create_connection((self.name, self.port), timeout=8))
                msg="{} is up on port {} with {}".format(self.name, self.port, self.connection)
                success=True
            else:
                if(self.ping()):
                    msg="{} is up on port {} with {}".format(self.name, self.port, self.connection)
                    success=True
                else:
                    msg="{} is not reachable".format(self.name)
                    if(self.priority=="high"):
                        try:
                            Gmail.email_alert("Server Down", msg, "anthony.calo@baruchmail.cuny.edu")
                        except Exception:
                            print("Couldn't send email becuase: {}".format(Exception))

        except socket.timeout:
            msg= "Server: {} timeout on port {}".format(self.name, self.port)
            if(self.priority=="high"):
                try:
                    Gmail.email_alert("Server Down", msg, "anthony.calo@baruchmail.cuny.edu")
                except Exception:
                    print("Couldn't send email becuase: {}".format(Exception))
        except Exception as e:
            msg="Server had this error: {}".format(e)
            if(self.priority=="high"):
                try:
                    Gmail.email_alert("Server Down", msg, "anthony.calo@baruchmail.cuny.edu")
                except Exception:
                    print("Couldn't send email becuase: {}".format(Exception))
        self.create_history(msg, success, now)
    
    def create_history(self, msg, success, time):
        for location in self.locations:
            with open(location[1], 'a') as f:
                f.write(msg)
                f.write("\n")
                f.write("Time: {}\nSuccess Status: {}\n".format(time, success))
                f.write("___________________________")
                f.write("\n")

    def ping(self):
        try:
            request=requests.get(self.name, timeout=5)
            return True
        except:
            return False
            pass

if __name__ == '__main__':
    servers= [
        Server("reddit.com", 80, "plain", "low"),
        Server("anthonycalo.com", 443, "ssl", "low"),
        Server("calochess.com", 80, "plain", "low"),
        Server("192.168.1.18", 80, "ping", "high")
    ]
    addServer = input("Would you like to add a server(yes or no): ")
    if(addServer.lower()=="yes"):
        servername=input("Enter server name: ")
        port =int(input("Enter port number: "))
        connection=input("Enter a type ping/ssl/plain: ")
        priority=input("Enter priority high/ low: ")
        servers.append(Server(servername, port, connection, priority))
        
    for server in servers:
        server.check_connection()








